src/phasorpy/_phasorpy.pyx

```diff
@@ -2041,72 +2041,66 @@ def _signal_denoise_vector(
 ###############################################################################
 # Filtering functions
 
-cdef float_t _quickselect(
-    float_t *arr,
-    ssize_t left,
-    ssize_t right,
-    const ssize_t k
-) noexcept nogil:
-    """Return k-th smallest array value using Quickselect algorithm."""
+
+cdef float_t _median(float_t *values, const ssize_t size) noexcept nogil:
+    """Return median of array values using Quickselect algorithm."""
     cdef:
-        ssize_t i, pivot_index, pivot_new_index
+        ssize_t i, pivot_index, pivot_index_new
+        ssize_t left = 0
+        ssize_t right = size - 1
+        ssize_t middle = size // 2
         float_t pivot_value, temp
 
+    if size % 2 == 0:
+        middle -= 1  # Quickselect sorts on right
+
     while left <= right:
         pivot_index = left + (right - left) // 2
-        pivot_value = arr[pivot_index]
-        temp = arr[pivot_index]
-        arr[pivot_index] = arr[right]
-        arr[right] = temp
-        pivot_new_index = left
+        pivot_value = values[pivot_index]
+        temp = values[pivot_index]
+        values[pivot_index] = values[right]
+        values[right] = temp
+        pivot_index_new = left
         for i in range(left, right):
-            if arr[i] < pivot_value:
-                temp = arr[i]
-                arr[i] = arr[pivot_new_index]
-                arr[pivot_new_index] = temp
-                pivot_new_index += 1
-        temp = arr[right]
-        arr[right] = arr[pivot_new_index]
-        arr[pivot_new_index] = temp
-
-        if pivot_new_index == k:
-            return arr[k]
-        if pivot_new_index < k:
-            left = pivot_new_index + 1
+            if values[i] < pivot_value:
+                temp = values[i]
+                values[i] = values[pivot_index_new]
+                values[pivot_index_new] = temp
+                pivot_index_new += 1
+        temp = values[right]
+        values[right] = values[pivot_index_new]
+        values[pivot_index_new] = temp
+
+        if pivot_index_new == middle:
+            if size % 2 == 0:
+                return (values[middle] + values[middle + 1]) / <float_t> 2.0
+            return values[middle]
+        if pivot_index_new < middle:
+            left = pivot_index_new + 1
         else:
-            right = pivot_new_index - 1
+            right = pivot_index_new - 1
 
-    return arr[k]
+    return values[middle]  # unreachable code?
```

Comment:
You are right Christoph, I think this line can be deleted since in all cases, the while loop will eventually find the median and return it.

Reply:
I'll leave it since formally the function has to return a value from all code paths.